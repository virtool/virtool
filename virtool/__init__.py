"""Application module.

TODO: Remove import when spaces are completely removed.
"""

import virtool.spaces.sql  # noqa: F401
