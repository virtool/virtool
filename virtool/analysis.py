LIST_PROJECTION = [
    "_id",
    "name",
    "algorithm",
    "timestamp",
    "ready",
    "job",
    "index_version",
    "user_id",
    "sample_id"
]


def processor(document):
    document = dict(document)
    document["analysis_id"] = document.pop("_id")

    return document
