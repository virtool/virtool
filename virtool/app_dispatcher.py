from copy import deepcopy

import virtool.handlers.utils


class Connection:

    def __init__(self, ws, session):
        self._ws = ws
        self.user_id = session.user_id
        self.groups = session.groups
        self.permissions = session.permissions

    async def send(self, message):
        await self._ws.send_json(message, dumps=virtool.handlers.utils.dumps)

    async def close(self):
        await self._ws.close()


class Dispatcher:

    def __init__(self):
        #: A dict of all active connections.
        self.connections = list()

    def add_connection(self, connection):
        """
        Add a connection to the dispatcher.

        """
        self.connections.append(connection)

    def remove_connection(self, connection):
        """
        Remove a connection from the dispatcher. Make sure it is closed first.

        :param connection: the connection to remove
        :type connection: :class:`.SocketHandler`

        """
        try:
            self.connections.remove(connection)
        except ValueError:
            pass

    async def dispatch(self, interface, operation, data, connections=None, conn_filter=None, conn_modifier=None,
                       writer=None):
        """
        Dispatch a ``message`` with a conserved format to a selection of active ``connections``.
        
        :param interface: the name of the interface the client should perform the operation on
        :type interface: str

        :param operation: a word used to tell the client what to do in response to the message
        :type operation: str
        
        :param data: the data the client will use
        :type data: dict

        :param connections: the connection(s) (:class:`.SocketHandler` objects) to dispatch the message to.
        :type connections: list

        :param conn_filter: filters the connections to which messages are written.
        :type conn_filter: callable

        :param conn_modifier: modifies the connection objects to which messages are written.
        :type conn_modifier: callable

        :param writer: modifies the written message based on the connection.
        :type writer: callable

        """
        message = {
            "operation": operation,
            "interface": interface,
            "data": data
        }

        # If the connections parameter was not set, dispatch the message to all authorized connections. Authorized
        # connections have assigned ``user_id`` properties.
        connections = connections or [conn for conn in self.connections if conn.user_id]

        if conn_filter:
            if not callable(conn_filter):
                raise TypeError("conn_filter must be callable")

            connections = [conn for conn in connections if conn_filter(conn)]

        if conn_modifier:
            if not callable(conn_modifier):
                raise TypeError("conn_modifier must be callable")

            for connection in connections:
                conn_modifier(connection)

        if writer:
            if not callable(writer):
                raise TypeError("writer must be callable")

            for connection in connections:
                await writer(connection, deepcopy(message))

            return

        for connection in connections:
            await connection.send(message)
