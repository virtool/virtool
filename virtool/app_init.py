import coloredlogs
import logging.handlers

from argparse import ArgumentParser


def get_args():

    parser = ArgumentParser()

    parser.add_argument(
        "-H", "--host",
        dest="host",
        default="localhost",
        help="hostname to listen on"
    )

    parser.add_argument(
        "-p", "--port",
        dest="port",
        default=9550,
        help="port to listen on"
    )

    parser.add_argument(
        "-v", "--verbose",
        dest="verbose",
        action="store_true",
        default=False,
        help="log debug messages"
    )

    parser.add_argument(
        "-P", "--write-pid",
        dest="write_pid",
        action="store_true",
        default=False,
        help="write a pidfile on start"
    )

    return parser.parse_args()


def configure(verbose=False):
    logging_level = logging.INFO if verbose else logging.DEBUG

    logging.captureWarnings(True)

    log_format = "%(asctime)-20s %(module)-11s %(levelname)-8s %(message)s"

    coloredlogs.install(
        level=logging_level,
        fmt=log_format
    )

    logger = logging.getLogger("virtool")

    handler = logging.handlers.RotatingFileHandler("virtool.log", maxBytes=1000000, backupCount=5)
    handler.setFormatter(logging.Formatter(log_format))

    logger.addHandler(handler)

    return logger
