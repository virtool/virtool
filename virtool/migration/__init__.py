"""Manage revisions in Alembic and other Virtool data.

* Virtool revisions are performed on PostgreSQL MongoDB, OpenFGA, and data files.
* Alembic revisions are for PostgreSQL only.

When should you use Virtool revisions for Postgres? When you need to make mass changes
to the database involving multiple transactions. For example, migrating OTU data from
MongoDB to PostgreSQL.

In Virtool, Alembic and non-Alembic revisions are managed together. If your newest
revision is an Alembic revision, and you create a new one. It will directly depend on
the previous Alembic revision having been applied.

# Guidelines

## MongoDB Transactions

In MongoDB, transactions fail frequently. We should avoid them for all but the smallest
operations.

It is not possible to guarantee that the revision will be recorded as complete even if
the ``upgrade()`` method succeeds. In this case, the revision will be run again next
time the migration is run.

Only write revisions for MongoDB that can be run again without causing problems and can
be re-run in the case that a previous run did not complete. You can do this by taking
the following precautions.

1. If you are adding a field, only add it to documents that do not have it.
2. If you are **not** using a transaction with ``update_many``, manually check that the
   update was successful. The ``update_many`` operation is not atomic.

   See existing revisions for how to do this.

## File Operations

Revisions containing file operations musts be written with the assumption that they
will fail and have to be re-run.

"""

from virtool.migration.cls import MigrationContext, MigrationError

__all__ = [
    "MigrationContext",
    "MigrationError",
]
