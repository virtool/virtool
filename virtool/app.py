import asyncio
import logging
import os
import sys

import aiohttp.web
import aiojobs
import aiojobs.aiohttp

import virtool.config
import virtool.db.core
import virtool.db.migrate
import virtool.db.utils
import virtool.dispatcher
import virtool.errors
import virtool.hmm.db
import virtool.http.accept
import virtool.http.auth
import virtool.http.csp
import virtool.http.errors
import virtool.http.proxy
import virtool.http.query
import virtool.jobs.runner
import virtool.logs
import virtool.references.db
import virtool.resources
import virtool.routes
import virtool.sentry
import virtool.settings.db
import virtool.settings.schema
import virtool.software.db
import virtool.startup
import virtool.utils
import virtool.version

logger = logging.getLogger(__name__)


async def on_shutdown(app: aiohttp.web.Application):
    """
    A function called when the app is shutting down.

    :param app: the app object
    :type app: :class:`aiohttp.web.Application`

    """
    logger.debug("Shutting down")

    try:
        await app["client"].close()
    except KeyError:
        pass

    try:
        await app["dispatcher"].close()
    except KeyError:
        pass

    try:
        app["executor"].shutdown(wait=True)
    except KeyError:
        pass

    try:
        app["process_executor"].shutdown(wait=True)
    except KeyError:
        pass

    scheduler = virtool.startup.get_scheduler_from_app(app)
    await scheduler.close()

    try:
        app["redis"].close()
        await app["redis"].wait_closed()
    except KeyError:
        pass


def create_app(config):
    """
    Creates the Virtool application.

    """
    middlewares = [
        virtool.http.auth.middleware,
        virtool.http.accept.middleware,
        virtool.http.csp.middleware,
        virtool.http.errors.middleware,
        virtool.http.proxy.middleware,
        virtool.http.query.middleware
    ]

    app = aiohttp.web.Application(middlewares=middlewares)

    app["config"] = config
    app["change_queue"] = asyncio.Queue()
    app["mode"] = "server"

    aiojobs.aiohttp.setup(app)

    app.on_startup.extend([
        virtool.startup.init_version,
        virtool.startup.init_events,
        virtool.startup.init_db,
        virtool.startup.init_dispatcher,
        virtool.startup.init_settings,
        virtool.startup.init_client_path,
        virtool.startup.init_http_client,
        virtool.startup.init_paths,
        virtool.startup.init_postgres,
        virtool.startup.init_routes,
        virtool.startup.init_executors,
        virtool.startup.init_tasks,
        virtool.startup.init_redis,
        virtool.startup.init_listen_for_changes,
        virtool.startup.init_sentry,
        virtool.startup.init_check_db,
        virtool.startup.init_resources,
        virtool.startup.init_job_interface,
        virtool.startup.init_refresh
    ])

    app.on_response_prepare.append(virtool.http.csp.on_prepare)
    app.on_shutdown.append(on_shutdown)

    return app


async def create_app_runner(app: aiohttp.web.Application, host: str, port: int) -> aiohttp.web.AppRunner:
    """
    Create an :class:`aiohttp.web.AppRunner` to allow customization of signal handlers.

    The basic :func:`aiohttp.web.run_app` sets up handlers for `SIGINT` and `SIGTERM` that can interfere with Virtool
    code such as that for restarting the server after software update. This custom runner allows handling of signals
    as well as restart and shutdown events from users.

    https://docs.aiohttp.org/en/stable/web_advanced.html#application-runners

    :param app: the application
    :param host: the host to listen on
    :param port: the port to listen on
    :return: a custom :class:`~aiohttp.web.AppRunner`

    """
    runner = aiohttp.web.AppRunner(app)

    await runner.setup()

    site = aiohttp.web.TCPSite(runner, host, port)

    try:
        await site.start()
    except OSError as err:
        if err.args[0] == 48:
            logger.fatal(f"Could not bind address {(host, port)}")
            sys.exit(1)

    logger.info(f"Listening at http://{host}:{port}")

    return runner


async def run_app(config):
    app = create_app(config)

    runner = await create_app_runner(
        app,
        config["host"],
        config["port"]
    )

    _, pending = await asyncio.wait(
        [wait_for_restart(runner, app["events"]), wait_for_shutdown(runner, app["events"])],
        return_when=asyncio.FIRST_COMPLETED
    )

    for task in pending:
        task.cancel()


async def wait_for_restart(runner: aiohttp.web.AppRunner, events: dict):
    """
    Wait for the shutdown event and restart if it is encountered.

    Restart is accomplished using :func:`os.execl` or :func:`os.execv` after cleaning up the `runner`.

    :param runner: the :class:`~aiohttp.web.AppRunner` returned by :func:`.create_app_runner`
    :param events: a dict containing the `restart` and `shutdown` :class:`~asyncio.Event` objects

    """
    await events["restart"].wait()
    await asyncio.sleep(0.5)
    await runner.cleanup()

    exe = sys.executable

    if exe.endswith("python") or "python3" in exe:
        os.execl(exe, exe, *sys.argv)
        return

    if exe.endswith("run"):
        os.execv(exe, sys.argv)
        return

    raise SystemError("Could not determine executable type")


async def wait_for_shutdown(runner: aiohttp.web.AppRunner, events: dict):
    """
    Wait for the shutdown event and terminate if it is encountered.

    :param runner: the :class:`~aiohttp.web.AppRunner` returned by :func:`.create_app_runner`
    :param events: a dict containing the `restart` and `shutdown` :class:`~asyncio.Event` objects

    """
    await events["shutdown"].wait()
    await asyncio.sleep(0.5)
    await runner.cleanup()
