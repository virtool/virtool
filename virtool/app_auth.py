import base64
from aiohttp import web

import virtool.user
import virtool.utils
import virtool.errors
from virtool.handlers.utils import bad_request


AUTHORIZATION_PROJECTION = ["user", "groups", "permissions"]


class Client:

    def __init__(self, ip, user_agent):
        # These attributes are assigned even when the session is not authorized.
        self.ip = ip
        self.user_agent = user_agent

        self.authorized = False
        self.user_id = None
        self.groups = None
        self.permissions = None
        self.is_api = False
        self.session_id = None

    def authorize(self, document, is_api):
        self.authorized = True
        self.user_id = document["user"]["id"]
        self.groups = document["groups"]
        self.permissions = document["permissions"]
        self.is_api = is_api

    def set_session_id(self, session_id):
        self.session_id = session_id


@web.middleware
async def middleware(req, handler):
    ip = req.transport.get_extra_info("peername")[0]
    user_agent = req.headers.get("User-Agent", None)

    req["client"] = Client(ip, user_agent)

    authorization = req.headers.get("AUTHORIZATION", None)

    can_use_api_key = req.path[0:4] == "/api" or req.path[0:7] == "/upload"

    # Try API key authorization if there was no session_id.
    if authorization and req.app["settings"].get("enable_api") and can_use_api_key:
        if authorization:
            try:
                user_id, key = decode_authorization(authorization)
            except virtool.errors.AuthError:
                return bad_request("Malformed Authorization header")

            document = await req.app["db"].keys.find_one({
                "_id": virtool.user.hash_api_key(key),
                "user.id": user_id
            }, AUTHORIZATION_PROJECTION)

            if not document:
                return bad_request("Invalid Authorization header")

            req["client"].authorize(document, True)

            return await handler(req)

    # Check if there is a session_id in cookies.
    session_id = req.cookies.get("session_id", None)

    document = await req.app["db"].sessions.find_one({
        "_id": session_id,
        "user_agent": user_agent
    })

    if document:
        user_id = None

        try:
            user_id = document["user"]["id"]
        except KeyError:
            pass

        if user_id:
            req["client"].authorize(document, False)
    else:
        session_id = virtool.utils.random_alphanumeric(128, mixed_case=True)

        await req.app["db"].sessions.insert_one({
            "_id": session_id,
            "ip": ip,
            "user_agent": user_agent
        })

    req["client"].set_session_id(session_id)

    response = await handler(req)

    response.set_cookie("session_id", session_id)

    return response


def decode_authorization(authorization):
    split = authorization.split(" ")

    if len(split) != 2 or split[0] != "Basic":
        raise virtool.errors.AuthError("Malformed authorization header")

    decoded = base64.b64decode(split[1]).decode("utf-8")

    return decoded.split(":")
