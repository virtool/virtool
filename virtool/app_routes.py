import logging

import virtool.api.account
import virtool.api.analyses
import virtool.api.caches
import virtool.api.downloads
import virtool.api.files
import virtool.api.genbank
import virtool.api.groups
import virtool.api.history
import virtool.api.hmm
import virtool.api.indexes
import virtool.api.jobs
import virtool.api.otus
import virtool.api.processes
import virtool.api.references
import virtool.api.root
import virtool.api.samples
import virtool.api.settings
import virtool.api.software
import virtool.api.subtractions
import virtool.api.uploads
import virtool.api.users
import virtool.api.websocket
import virtool.http.auth
import virtool.utils
import virtool.http.auth

logger = logging.getLogger(__name__)

INDEX_PATHS = [
    "/",
    r"/account{suffix:.*}",
    r"/administration{suffix:.*}",
    r"/home{suffix:.*}",
    r"/hmm{suffix:.*}",
    r"/jobs{suffix:.*}",
    r"/otus{suffix:.*}",
    r"/refs{suffix:.*}",
    r"/samples{suffix:.*}",
    r"/subtraction{suffix:.*}"
]


def setup_routes(app):
    for path in INDEX_PATHS:
        app.router.add_get(path, virtool.http.auth.index_handler)

    app.router.add_get("/ws", virtool.api.websocket.root)
    app.router.add_get("/login", virtool.http.auth.login_get_handler)
    app.router.add_post("/login", virtool.http.auth.login_post_handler)
    app.router.add_get("/reset", virtool.http.auth.reset_get_handler)
    app.router.add_post("/reset", virtool.http.auth.reset_post_handler)

    app.router.add_routes(virtool.api.account.routes)
    app.router.add_routes(virtool.api.analyses.routes)
    app.router.add_routes(virtool.api.caches.routes)
    app.router.add_routes(virtool.api.downloads.routes)
    app.router.add_routes(virtool.api.files.routes)
    app.router.add_routes(virtool.api.genbank.routes)
    app.router.add_routes(virtool.api.groups.routes)
    app.router.add_routes(virtool.api.history.routes)
    app.router.add_routes(virtool.api.hmm.routes)
    app.router.add_routes(virtool.api.indexes.routes)
    app.router.add_routes(virtool.api.jobs.routes)
    app.router.add_routes(virtool.api.otus.routes)
    app.router.add_routes(virtool.api.processes.routes)
    app.router.add_routes(virtool.api.references.routes)
    app.router.add_routes(virtool.api.root.routes)
    app.router.add_routes(virtool.api.samples.routes)
    app.router.add_routes(virtool.api.settings.routes)
    app.router.add_routes(virtool.api.software.routes)
    app.router.add_routes(virtool.api.subtractions.routes)
    app.router.add_routes(virtool.api.uploads.routes)
    app.router.add_routes(virtool.api.users.routes)
