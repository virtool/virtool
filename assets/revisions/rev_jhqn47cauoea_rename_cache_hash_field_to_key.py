"""
Rename cache hash field to key

Revision ID: jhqn47cauoea
Date: 2022-06-09 22:12:49.222586

"""
import asyncio

import arrow

from virtool.migration import MigrationContext, MigrationError

# Revision identifiers.
name = "Rename cache hash field to key"
created_at = arrow.get("2022-06-09 22:12:49.222586")
revision_id = "jhqn47cauoea"

alembic_down_revision = None
virtool_down_revision = "1keyha5n6l0j"


async def upgrade(ctx: MigrationContext):
    await ctx.mongo.caches.update_many({}, {"$rename": {"hash": "key"}})

    if await ctx.mongo.caches.count_documents({"hash": {"$exists": True}}):
        raise MigrationError("Some caches still have a hash field")


async def test_upgrade(ctx: MigrationContext, snapshot):
    await asyncio.gather(
        ctx.mongo.caches.insert_many(
            [
                {
                    "_id": "foo",
                    "hash": "a97439e170adc4365c5b92bd2c148ed57d75e566",
                    "sample": {"id": "abc"},
                },
                {
                    "_id": "bar",
                    "hash": "d7fh3ee170adc4365c5b92bd2c1f3fd5745te566",
                    "sample": {"id": "dfg"},
                },
            ]
        )
    )

    await upgrade(ctx)

    assert await ctx.mongo.caches.find().to_list(None) == snapshot
