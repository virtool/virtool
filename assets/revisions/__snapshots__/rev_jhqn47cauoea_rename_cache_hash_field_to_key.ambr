# name: test_upgrade[uvloop]
  <class 'list'> [
    <class 'dict'> {
      '_id': 'foo',
      'key': 'a97439e170adc4365c5b92bd2c148ed57d75e566',
      'sample': <class 'dict'> {
        'id': 'abc',
      },
    },
    <class 'dict'> {
      '_id': 'bar',
      'key': 'd7fh3ee170adc4365c5b92bd2c1f3fd5745te566',
      'sample': <class 'dict'> {
        'id': 'dfg',
      },
    },
  ]
---
